### ISSUE TITLE
Memory leaks

### ISSUE THREAD

**Issue by Ixiodor** [celery/celery#6132]
# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [x] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [ ] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [ ] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [ ] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**: celery==4.4.2

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
[2020-05-31 23:28:34,434: INFO/MainProcess] Connected to amqp://remote_worker:**@127.0.0.1:5672//
[2020-05-31 23:28:34,453: INFO/MainProcess] mingle: searching for neighbors
[2020-05-31 23:28:35,487: INFO/MainProcess] mingle: all alone
[2020-05-31 23:28:35,528: WARNING/MainProcess] /home/ubuntu/.local/lib/python3.7/site-packages/celery/fixups/django.py:203: UserWarning: Using settings.DEBUG leads to a memory
            leak, never use this setting in production environments!
  leak, never use this setting in production environments!''')
[2020-05-31 23:28:35,529: INFO/MainProcess] celery@testroom ready.
[2020-05-31 23:28:47,351: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  
[2020-05-31 23:28:47,689: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:28:47,690: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:28:47,690: WARNING/ForkPoolWorker-1] retry
[2020-05-31 23:28:47,721: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  ETA:[2020-05-31 23:28:57.692348+00:00] 
[2020-05-31 23:28:47,722: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  ETA:[2020-05-31 23:28:57.716321+00:00] 
[2020-05-31 23:28:47,777: INFO/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] retry: Retry in 10s: Retry(Retry(...), Exception('i have filled now'), 10)
[2020-05-31 23:28:57,999: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:28:57,999: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:28:58,000: WARNING/ForkPoolWorker-1] ended
[2020-05-31 23:28:58,000: WARNING/ForkPoolWorker-1] BYEEEEE
[2020-05-31 23:28:58,062: INFO/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] succeeded in 0.34440315900428686s: None
[2020-05-31 23:28:58,301: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:28:58,302: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:28:58,302: WARNING/ForkPoolWorker-1] retry
[2020-05-31 23:28:58,304: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  ETA:[2020-05-31 23:29:08.303091+00:00] 
[2020-05-31 23:28:58,307: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  ETA:[2020-05-31 23:29:08.306141+00:00] 
[2020-05-31 23:28:58,368: INFO/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] retry: Retry in 10s: Retry(Retry(...), Exception('i have filled now'), 10)
[2020-05-31 23:29:08,572: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:29:08,572: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:29:08,573: WARNING/ForkPoolWorker-1] ended
[2020-05-31 23:29:08,573: WARNING/ForkPoolWorker-1] BYEEEEE
[2020-05-31 23:29:08,633: INFO/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] succeeded in 0.3256059319974156s: None
[2020-05-31 23:29:08,872: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:29:08,872: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:29:08,873: WARNING/ForkPoolWorker-1] retry
[2020-05-31 23:29:08,875: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  ETA:[2020-05-31 23:29:18.873799+00:00] 
[2020-05-31 23:29:08,878: WARNING/ForkPoolWorker-1] BYEEEEE
[2020-05-31 23:29:08,880: INFO/MainProcess] Received task: api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906]  ETA:[2020-05-31 23:29:18.877550+00:00] 
[2020-05-31 23:29:08,940: INFO/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] retry: Retry in 10s: Retry(Retry(...), Exception('i have filled now'), 10)
[2020-05-31 23:29:19,144: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:29:19,144: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:29:19,145: WARNING/ForkPoolWorker-1] ended
[2020-05-31 23:29:19,145: WARNING/ForkPoolWorker-1] BYEEEEE
[2020-05-31 23:29:19,205: INFO/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] succeeded in 0.326258520995907s: None
[2020-05-31 23:29:19,444: WARNING/ForkPoolWorker-1] started
[2020-05-31 23:29:19,444: WARNING/ForkPoolWorker-1] HELLOOOOO
[2020-05-31 23:29:19,445: WARNING/ForkPoolWorker-1] retry
[2020-05-31 23:29:19,505: ERROR/ForkPoolWorker-1] Task api_v3.tests.execute[e97d93b5-b0e5-4b87-96ab-1aab66119906] raised unexpected: Exception('i have filled now')
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/app/trace.py", line 385, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/app/trace.py", line 650, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/app/base.py", line 500, in run
    raise task.retry(exc=exc, **retry_kwargs)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/app/task.py", line 704, in retry
    raise_with_context(exc)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/app/base.py", line 487, in run
    return task._orig_run(*args, **kwargs)
  File "/var/www/django_projects/earthalytics-api/api_v3/tests.py", line 26, in execute
    self.retry(exc=Exception("i have filled now"), args=[param_a, param_b], kwargs=kwargs)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/app/task.py", line 704, in retry
    raise_with_context(exc)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/celery/utils/serialization.py", line 288, in raise_with_context
    _raise_with_context(exc, exc_info[1])
  File "<string>", line 1, in _raise_with_context
Exception: i have filled now

```

</p>
</details>

# Steps to Reproduce
Make a class with a destructor.
Make a celery task.
Instantiate the class in the celery task.
Make the celery task raise an Exception.
See how destructor dosn't get called.(Gets called if task ends in success)

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
class WelcomeClass():
    def __init__(self):
        print("HELLOOOOO")

    def __del__(self):
        print("BYEEEEE")


@app_cluster.task(bind=True, autoretry_for=(Exception,), max_retries=3,
                  default_retry_delay=10)
def execute(self, param_a, param_b=None, **kwargs):
    print("started")
    jb = WelcomeClass()
    if param_b is None:
        param_b = "filled"
        print("retry")
        self.retry(exc=Exception("i have filled now"), args=[param_a, param_b], kwargs=kwargs)
    print("ended")

def test_celery(self):
    sig = execute.si("something")
    t = sig.delay()
    t = 0
```

</p>
</details>

# Expected Behavior

I expect the destructor of a class begin called everytime a celery task succeed/fail/raise exception etc
# Actual Behavior
 Destrucutor gets called only when task succeed.

**Comment by auvipy** [comment]
your report is not complete.

**Comment by Ixiodor** [comment]
Updated.
As you think it's good, I'll copy paste this for the second BUG that need the same code.

**Comment by mchataigner** [comment]
@Ixiodor, could you come back at us if this is fixed as well as you mentioned here https://github.com/celery/celery/pull/6137#issuecomment-637141700

**Comment by Ixiodor** [comment]
I'll verify if it's all ok and I'll give feedback.
My code needs to play around class destructor.

**Comment by brianwawok** [comment]
Confirmed 4.4.3 brought down prod due to a memory leak for me also

**Comment by auvipy** [comment]
> Confirmed 4.4.3 brought down prod due to a memory leak for me also

can you try the master? there are some new fixes on master. gonna release a new version in this week

**Comment by brianwawok** [comment]
Yes I can run an experiment with master later today.

Basically my symptoms are that one of my worker pools (with the highest memory limit), would slowly gain memory, until the OOM killer would kill a worker, but the worker would hang rather than die - so never get respawned. So after 30 minutes the end result is 100% of workers are "hung".

I upgraded celery, kombu, and AMPQ in the same push - so not 100% sure who was to blame. Was on Celery 4.4.2 which (if this bug report is for 4.4.2) I worry that maybe it wasn't celery, and it was down the kombu, AMPQ line.

**Comment by Ixiodor** [comment]
do your tasks use retry os has some retry mechanics?(autoretry_for)

**Comment by auvipy** [comment]
can you try fil profiler?

**Comment by brianwawok** [comment]
Yes, my tasks heavily abuse auto-retry for on a list of exceptions. Some have decent sized payloads (1-10 MB) that I am aiming to refactor out but haven't get.

I do log individual task memory to a DB, but it doesn't go down to the level of fil profiler. Would have to read into how it works.

**Comment by Ixiodor** [comment]
Ok, it's this.
Retry is heavly compromised

**Comment by Ixiodor** [comment]
Allocation/Deallocation seems working fine in my environment. I have to wait a bit more for a production test, so, i would to wait news from @brianwawok too.

**Comment by heedong-jung** [comment]
this could be fixed in PR #6024. 
Does it happen with the most up-to-date code?

**Comment by auvipy** [comment]
we got a new celery release in pypi
